vpr/src/route/clock_connection_builders.cpp

```diff
@@ -95,35 +93,38 @@ int RoutingToClockConnection::create_virtual_clock_network_sink_node(
     int x,
     int y) {
     auto& device_ctx = g_vpr_ctx.mutable_device();
-    auto& rr_nodes = device_ctx.rr_nodes;
-    rr_nodes.emplace_back();
-    auto node_index = rr_nodes.size() - 1;
+    auto& rr_graph = device_ctx.rr_nodes;
+    auto& node_lookup = device_ctx.rr_graph_builder.node_lookup();
+    rr_graph.emplace_back();
+    RRNodeId node_index = RRNodeId(rr_graph.size() - 1);
 
     //Determine the a valid PTC
-    std::vector<int> nodes_at_loc;
-    get_rr_node_indices(device_ctx.rr_node_indices,
-                        x, y,
-                        SINK,
-                        &nodes_at_loc);
+    std::vector<RRNodeId> nodes_at_loc = node_lookup.find_sink_nodes(x, y);
 
     int max_ptc = 0;
-    for (int inode : nodes_at_loc) {
-        max_ptc = std::max<int>(max_ptc, device_ctx.rr_nodes[inode].ptc_num());
+    for (RRNodeId inode : nodes_at_loc) {
+        max_ptc = std::max<int>(max_ptc, rr_graph.node_ptc_num(inode));
     }
     int ptc = max_ptc + 1;
 
-    rr_nodes[node_index].set_ptc_num(ptc);
-    rr_nodes[node_index].set_coordinates(x, y, x, y);
-    rr_nodes[node_index].set_capacity(1);
-    rr_nodes[node_index].set_cost_index(SINK_COST_INDEX);
-    rr_nodes[node_index].set_type(SINK);
+    rr_graph.set_node_ptc_num(node_index, ptc);
+    rr_graph.set_node_coordinates(node_index, x, y, x, y);
+    rr_graph.set_node_capacity(node_index, 1);
+    rr_graph.set_node_cost_index(node_index, SINK_COST_INDEX);
+    rr_graph.set_node_type(node_index, SINK);
     float R = 0.;
     float C = 0.;
-    rr_nodes[node_index].set_rc_index(find_create_rr_rc_data(R, C));
+    rr_graph.set_node_rc_index(node_index, find_create_rr_rc_data(R, C));
 
-    add_to_rr_node_indices(device_ctx.rr_node_indices, rr_nodes, node_index);
+    // Use a generic way when adding nodes to lookup.
+    // However, since the SINK node has the same xhigh/xlow as well as yhigh/ylow, we can probably use a shortcut
+    for (int ix = rr_graph.node_xlow(node_index); ix <= rr_graph.node_xhigh(node_index); ++ix) {
+        for (int iy = rr_graph.node_ylow(node_index); iy <= rr_graph.node_yhigh(node_index); ++iy) {
+            node_lookup.add_node(node_index, ix, iy, rr_graph.node_type(node_index), rr_graph.node_ptc_num(node_index), SIDES[0]);
```

Comment:
Looks like we use SIDES[0] to insert the sink node, and look for it later on the TOP side.  Hence should we use TOP consistently or SIDES[0] consistently to make sure this is always in alignment?  I assume SIDES[0] == TOP so this all works, but it still seems dangerous as if anyone ever edits the order of SIDES we'd subtly break code written this way.

Reply:
Agree. We should not expose the side argument to developers because they may feel confused.
You are right on the SIDE[0] which is actually the TOP side. This convention has been there for a while (I do not know who introduced it). I am o.k. to change it if necessary.

Actually, the side argument for non-IPIN/OPIN nodes will be overwritten inside the API.
As a result, developers cannot mess up the side by giving any invalid values.

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/c06800fb49d2732a0fcedbb2f542613fedd6a330/vpr/src/device/rr_spatial_lookup.cpp#L23-L29

However, I think it is not a good idea to force the developers to give a dummy side when using the API. I modified the API so that the side option will have a default value (SIDE[0]) if not specified. In client functions, I remove all the usage of the SIDE[0] for non-IPIN/OPIN nodes.